grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java

```diff
@@ -745,6 +806,68 @@ private ImmutableList.Builder<ServerInterceptor> interceptors() {
         return interceptors;
     }
 
+    private void collectDecorators(Class<?> clazz, @Nullable String path,
+                                   ServerServiceDefinition serverServiceDefinition) {
+        final String serviceName =
+                path != null ? path : '/' + serverServiceDefinition.getServiceDescriptor().getName();
+        // In gRPC, A method name is unique.
+        final Map<String, List<Method>> methods = new HashMap<>();
+        for (Method method : getAllMethods(clazz, withModifier(Modifier.PUBLIC))) {
+            final String methodName = method.getName();
+            if (methods.containsKey(methodName)) {
+                methods.get(methodName).add(method);
+            } else {
+                methods.put(methodName, Lists.newArrayList(method));
+            }
+        }
+        for (final ServerMethodDefinition<?, ?> serverMethodDefinition : serverServiceDefinition.getMethods()) {
+            final String targetMethodName = serverMethodDefinition.getMethodDescriptor().getBareMethodName();
+            if (targetMethodName == null) {
+                continue;
+            }
+            final String methodName = targetMethodName.substring(0, 1)
+                                                      .toLowerCase() + targetMethodName.substring(1);
+            final List<Method> foundMethods = methods.get(methodName);
+            if (foundMethods == null) {
+                continue;
+            }
+            for (Method method : foundMethods) {
+                final List<DecoratorAndOrder> decorators = DecoratorAnnotationUtil.collectDecorators(clazz,
+                                                                                                     method);
+                if (!decorators.isEmpty()) {
+                    String key = serviceName + '/' + targetMethodName;
+                    if (pathToDecorators.containsKey(key)) {
+                        if (decorators.size() > pathToDecorators.get(key).size()) {
+                            pathToDecorators.put(key, decorators);
+                        }
```

Comment:
@ta7uw Could you explain why we need this condition?

Reply:
@ikhoon 
Consider the folliowing hierarchical service class,  the key for `UnaryCall` is `/armeria.grpc.testing.TestService/UnaryCall` and the keys generated from each of `ThirdTestServiceImpl`  and `AbstractThirdTestServiceImpl` is the same.
We should register the decorators of `ThirdTestServiceImpl`, so the number of decorators is used to determine which class is subclasses at each path.

```java
    @Decorator(Decorator1.class)
    private abstract static class AbstractThirdTestServiceImpl extends TestServiceImplBase {

        @Override
        @Decorator(Decorator2.class)
        public void unaryCall(SimpleRequest request, StreamObserver<SimpleResponse> responseObserver) {
            responseObserver.onNext(SimpleResponse.newBuilder()
                                                  .setUsername("test user")
                                                  .build());
            responseObserver.onCompleted();
        }
    }

    private static class ThirdTestServiceImpl extends AbstractThirdTestServiceImpl {

        @Override
        @Decorator(Decorator3.class)
        public void emptyCall(Empty request, StreamObserver<Empty> responseObserver) {
            responseObserver.onNext(Empty.getDefaultInstance());
            responseObserver.onCompleted();
        }
    }
```

Reply:
Thanks for the detailed explanation.

>  We should register the decorators of ThirdTestServiceImpl, so the number of decorators is used to determine which class is subclasses at each path.

It might be abnormal to pick up a super method because it has many more decorators. 
Based on the implementation of `ReflectionUtils.getAllMethods()`, it recursively collects methods from the bottom type so I can assume the order of method are sorted. A lower type will take precedence.
Should we modify the following code to get only the lowest implemented method?
https://github.com/line/armeria/blob/aca2ba6e5337c8d58adf48887453f735d664e5df/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java#L815-L822

Reply:
> Based on the implementation of ReflectionUtils.getAllMethods(), it recursively collects methods from the bottom type so I can assume the order of method are sorted

I think these methods are collected into HashSet, so the order of methods is not sorted

Reply:
> Based on the implementation of ReflectionUtils.getAllMethods(), it recursively collects methods from the bottom type so I can assume the order of method are sorted.

I think that methods are collected into HashSet in `ReflectionUtils`, so the order of methods is not sorted.
Should we copy methods in `ReflectionUtils` and refactor them to return collections as an ArrayList?

Reply:
> so the order of methods is not sorted.

Good point. I missed the detail.

>  Should we copy methods in ReflectionUtils and refactor them to return collections as an ArrayList?

Sounds good!